packages/core/src/modules/oob/OutOfBandModule.ts

```diff
@@ -0,0 +1,297 @@
+   *
+   * @param outOfBandMessage
+   */
+  public async receiveMessage(
+    outOfBandMessage: OutOfBandMessage,
+    config: ReceiveOutOfBandMessageConfig
+  ): Promise<ConnectionRecord | undefined> {
+    const { handshakeProtocols, services } = outOfBandMessage
+    const { autoAccept, reuse } = config
+    const messages = outOfBandMessage.getRequests()
+
+    if ((!handshakeProtocols || handshakeProtocols.length === 0) && (!messages || messages?.length === 0)) {
+      throw new AriesFrameworkError(
+        'One or both of handshake_protocols and requests~attach MUST be included in the message.'
+      )
+    }
+
+    if (handshakeProtocols) {
+      if (!this.areHandshakeProtocolsSupported(handshakeProtocols)) {
+        throw new AriesFrameworkError('Handshake protocols are not supported.')
+      }
+
+      let connectionRecord: ConnectionRecord
+      const existingConnection = await this.findExistingConnection(services)
+      if (existingConnection) {
+        this.logger.debug('Connection already exists.', { connectionId: existingConnection.id })
+        if (reuse) {
+          this.logger.debug('Reuse is enabled. Reusing an existing connection.')
+          connectionRecord = existingConnection
+          if (!messages) {
+            this.logger.debug('Out of band message does not contain any request messages.')
+            await this.sendReuse(outOfBandMessage, connectionRecord)
+          }
+        } else {
+          this.logger.debug('Reuse is disabled. Creating a new connection.')
+          connectionRecord = await this.createConnection(outOfBandMessage, autoAccept)
+        }
+      } else {
+        this.logger.debug('Connection does not exists. Creating a new connection.')
+        connectionRecord = await this.createConnection(outOfBandMessage, autoAccept)
+      }
+
+      if (messages) {
+        this.logger.debug('Out of band message contains request messages.')
+        if (!connectionRecord.isReady) {
+          // Wait until the connecion is ready and then pass the messages to the agent for further processing
+          this.connectionsModule
+            .returnWhenIsConnected(connectionRecord.id)
+            .then((c) => {
+              const connectionServices = c.theirDidDoc?.didCommServices
+              this.emitMessages(connectionServices, messages)
+            })
+            .catch((error) => {
+              if (error instanceof EmptyError) {
+                this.logger.warn(
+                  `Agent unsubscribed before connection got into ${ConnectionState.Complete} state`,
+                  error
+                )
+              } else {
+                this.logger.error('Promise waiting for the connection to be complete failed.', error)
+              }
+            })
+        } else {
+          const connectionServices = connectionRecord.theirDidDoc?.didCommServices
+          this.emitMessages(connectionServices, messages)
```

Comment:
I know you put this in the TODO, but we shouldn't emit all messages. I think it is important to fix this as it is not an incomplete way to handle it, but an incorrect way to handle it.

Reply:
Updated. It takes only the first supported message and throws an error if there is no message supported.

I haven't found a precise definition in the spec if we should allow only the "request" type of message from each protocol. What do you think?